introspection-engine/connectors/sql-introspection-connector/src/sanitize_datamodel_names.rs

```diff
@@ -1,44 +1,75 @@
 
             if let FieldType::Relation(info) = &mut field.field_type {
                 info.name = sanitize_name(info.name.clone()).0;
                 info.to = sanitize_name(info.to.clone()).0;
-                info.to_fields = info.to_fields.iter().map(|f| sanitize_name(f.clone()).0).collect();
+                info.to_fields = info
+                    .to_fields
+                    .iter()
+                    .map(|f: &std::string::String| sanitize_name(f.clone()).0)
+                    .collect();
+            }
+
+            if let FieldType::Enum(enum_name) = &mut field.field_type {
+                let (sanitized_enum_name, enum_db_name) = if *enum_name == format!("{}_{}", model.name, field.name) {
+                    //MySql
+                    if model_db_name.is_none() && field_db_name.is_none() {
+                        (enum_name.clone(), None)
+                    } else {
+                        (
+                            format!("{}_{}", sanitized_model_name, sanitized_field_name),
+                            Some(enum_name.clone()),
+                        )
+                    }
+                } else {
+                    sanitize_name(enum_name.clone())
+                };
+
+                if let Some(old_name) = enum_db_name {
+                    enum_renames.insert(old_name.clone(), (sanitized_enum_name.clone(), Some(old_name.clone())));
+                };
+
+                *enum_name = sanitized_enum_name;
+            }
+
+            field.name = sanitized_field_name;
+
+            if field.database_names.is_empty() {
+                field.database_names = field_db_name.map(|db| vec![db]).unwrap_or(vec![]);
             }
         }
 
         for index in &mut model.indices {
             index.fields = index.fields.iter().map(|f| sanitize_name(f.clone()).0).collect();
         }
+
+        model.name = sanitized_model_name;
+        model.database_name = model_db_name;
+    }
+
+    for enm in &mut datamodel.enums {
+        if let Some((sanitized_enum_name, enum_db_name)) = enum_renames.get(&enm.name) {
+            enm.name = sanitized_enum_name.to_owned();
+            enm.database_name = enum_db_name.to_owned();
+        } else {
+            let (sanitized_enum_name, enum_db_name) = sanitize_name(enm.name.clone());
+            enm.name = sanitized_enum_name.to_owned();
+            enm.database_name = enum_db_name.to_owned();
+        }
     }
 
-    //    todo enums are a bit more complicated and not fully specced, lets do this separately.
-    //    We also need to map the actual valid values before printing them
-    //    for enm in &mut datamodel.enums {
-    //        let (sanitized_name, db_name) = sanitize_name(enm.name.clone());
-    //
-    //        enm.name = sanitized_name;
-    //        enm.database_name = db_name.map(|db| Single(db));
-    //    }
+    // todo: do a pass over all modified names and deduplicate them
```

Comment:
it's hard for me to completely wrap my head around this function, but isn't that a problem that can be solved by using a `BTreeSet` or a `HashSet`?

Reply:
Sorry, now I get it, it's not that simple.